appsrc/dcfile/dcfile.man
.TH DCFILE 1 "6 March 2007" "DICOM PS3" "DICOM PS3 - Describe DICOM file content"
.SH NAME
dcfile \- ACR/NEMA DICOM PS3 ... DICOM PS3 - Describe DICOM file content
.SH SYNOPSIS
.HP 10
.B dcfile
[
.B \-b|brief
]
[
.B \-filename
]
.so man1/optin.so
.SH DESCRIPTION
.LP
.B dcfile
reads the named dicom or acr-nema input file and describes the encoding
method of the file, ie. what transfer syntax is in use, and whether or
not a Part 10 style meta-information header is present.
.SH OPTIONS
The description goes to standard error.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-b|brief
.RS
Display a single line brief summary.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dcfile dicomdir.
.RE
Meta: UID               1.2.840.10008.1.2.1
.RE
Meta: Description       "Explicit VR Little Endian"
.RE
Meta: ByteOrder Little
.RE
Meta: VR                Explicit
.RE
Meta: Encapsulated      No
.RE
Data: UID               1.2.840.10008.1.2.1
.RE
Data: Description       "Explicit VR Little Endian"
.RE
Data: ByteOrder Little
.RE
Data: VR                Explicit
.RE
Data: Encapsulated      No
.RE
\ 
.RE
% dcfile expmeta.dc3
.RE
Meta: UID               1.2.840.10008.1.2.1
.RE
Meta: Description       "Explicit VR Little Endian"
.RE
Meta: ByteOrder Little
.RE
Meta: VR                Explicit
.RE
Meta: Encapsulated      No
.RE
Data: UID               1.2.840.10008.1.2
.RE
Data: Description       "Implicit VR Little Endian"
.RE
Data: ByteOrder Little
.RE
Data: VR                Implicit
.RE
Data: Encapsulated      No
.RE
\ 
.RE
% dcfile NM.1
.RE
Data: UID               1.2.840.10008.1.2
.RE
Data: Description       "Implicit VR Little Endian"
.RE
Data: ByteOrder Little
.RE
Data: VR                Implicit
.RE
Data: Encapsulated      No
.RE
\ 
.RE
% dcfile test.spi
.RE
Data: ByteOrder Big
.RE
Data: VR                Implicit
.RE
Data: Encapsulated      No
.RE
\ 
.RE
% dcfile -b evrlewm.dcm
.RE
DICOM file, with metaheader, Explicit VR Little Endian
.RE
\ 
.RE
% dcfile -b evrlenom.dcm
.RE
DICOM file, no metaheader, Explicit VR Little Endian
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR file(1) ,
.BR dcdump(1) ,
.BR dcintro(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dcfile.cc
#include "basetype.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "dcstream.h"
#include "attr.h"
#include "attrnew.h"
#include "attrval.h"
#include "elmconst.h"
#include "transynd.h"

static bool
isOtherByteOrWordVR(const char *vr)
{
	return vr && vr[0]=='O' && (vr[1]=='B' || vr[1]=='W' || vr[1]=='X');
}

static bool
isSequenceVR(const char *vr)
{
	return vr && vr[0]=='S' &&  vr[1]=='Q';
}

static bool
isUnknownVR(const char *vr)
{
	return vr && vr[0]=='U' &&  vr[1]=='N';
}

static TransferSyntax *
readMetaHeaderTillTransferSyntax(DicomInputStream &stream)
{
	Uint32 n=0;
	Uint32 length=0xffffffff;
	while (stream.peek() != EOF && (length == 0xffffffff || n < length)) {
		Tag tag;
		stream >> tag;
		if (stream.fail()) {
			cerr << EMsgDC(TagReadFailed) << endl;
			return NULL;
		}
		n+=4;

		char vr[3];
		if (stream.getTransferSyntaxInUse()->isExplicitVR()) {
			stream.read(vr,2);
			if (stream.fail()) {
				cerr << EMsgDC(VRReadFailed) << endl;
				return NULL;
			}
			n+=2;
			vr[2]=0;
		}
		else {
			vr[0]=0;
		}

		Uint32 vl;
		if (!stream.getTransferSyntaxInUse()->isExplicitVR()) {
			stream >> vl;
			n+=4;
		}
		else if (isOtherByteOrWordVR(vr) || isSequenceVR(vr) || isUnknownVR(vr)) {
			(void)stream.read16();	// "Reserved"
			stream >> vl;
			n+=6;
		}
		else {
			vl=stream.read16();
			n+=2;
		}
		if (stream.fail()) {
			cerr << EMsgDC(VLReadFailed) << endl;
			return NULL;
		}

		if (tag == TagFromName(FileMetaInformationGroupLength)) {
			if (vl == 4) {
				length=stream.read32();
				n+=vl;
			}
			else {
				cerr << EMsgDC(UnexpectedVL) << endl;
				return NULL;
			}
		}
		else if (tag == TagFromName(TransferSyntaxUID)) {
			if (!stream.getTransferSyntaxInUse()->isExplicitVR()) {
				vr[0]='U'; vr[1]='I'; vr[2]=0;
			}
			Attribute *a=newAttribute(vr,tag);
			if (a) {
				a->read(stream,vl);
				if (stream.fail()) {
					cerr << EMsgDC(VLReadFailed) << endl;
					return NULL;
				}
				else {
					n+=vl;
					TransferSyntax *dts=new TransferSyntax(AttributeValue(a));
					Assert(dts);
					return dts;
				}
			}
			else {
				cerr << EMsgDC(UnrecognizedVR) << endl;
				return NULL;
			}
		}
		else if (!tag.isMetaheaderGroup()) {
			cerr << EMsgDC(PrematureEndOfMetaHeader) << endl;
			return NULL;
		}
		else if (vl == 0xffffffff) {
			cerr << EMsgDC(UnexpectedVL) << endl;
			return NULL;
		}
		else {
			stream.seekg(vl,ios::cur);
			if (stream.fail()) {
				cerr << EMsgDC(SeekFailed) << endl;
				return NULL;
			}
			n+=vl;
		}
	}
	cerr << EMsgDC(PrematureEndOfMetaHeader) << endl;
	return NULL;
}

int
main(int argc,char **argv)
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);

	bool brief=options.get("brief") || options.get("b");
	bool showfilename=options.get("filename");

	dicom_input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);

	cerr << dicom_input_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();

	if (!dicom_input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< " [-b|-brief]"
			<< " [-filename]"
			<< " [" << MMsgDC(InputFile) << "]"
			<< " <" << MMsgDC(InputFile)
			<< endl;
		return 1;
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);

	bool success;

	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		cerr << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	if (din.haveMetaHeader()) {
		TransferSyntax *mts=din.getTransferSyntaxToReadMetaHeader();
		if (mts) {
			if (!brief) {
				dumpTransferSyntax(mts,"Meta: ");
			}
			TransferSyntax *dts = readMetaHeaderTillTransferSyntax(din);
			if (dts) {
				if (brief) {
					cerr << "DICOM file, with metaheader, " << dts->getDescription() << endl;	// do not worry if mts is not EVRLE
				}
				else {
					dumpTransferSyntax(dts,"Data: ");
				}
				success = true;
			}
			else {
				success=false;
			}
		}
		else
			success=false;
	}
	else {
		TransferSyntax *dts=din.getTransferSyntaxToReadDataSet();
		if (dts) {
			if (brief) {
				cerr << "DICOM file, no metaheader, " << dts->getDescription() << endl;
			}
			else {
				dumpTransferSyntax(dts,"Data: ");
			}
		}
		if (din.isSwapped32Big() && !brief) {
			cerr << "Data: Swapped32Big\tYes" << endl;
		}
		success=true;
	}

	return success ? 0 : 1;
}

